std/concurrency.d

```diff
@@ -342,17 +342,17 @@ public:
      * that a Tid executed in the future will have the same toString() output
      * as another Tid that has already terminated.
      */
-    void toString(scope void delegate(const(char)[]) sink)
+    void toString(W)(ref W w) const
```

Comment:
What is the point of templating this function if you are going to put the @system line in a trusted block? You can simply leave the function signature as is and simply add @safe then.

Reply:
> What is the point of templating this function if you are going to put the @System line in a trusted block?

Using template functions is required in [std.format.formatValue](https://dlang.org/phobos/std_format.html#.formatValue) that is used in `formattedWrite`.
The current signature is still available but not recommended because it prevents from inferring attributes such as safe.

In the case of this request, I first tried to solve it by adding safe attribute to the current signature as you suggested.
However, the compiler said that "safe function `toSTring` cannot call a system function `formattedWrite`" and I finally found that the problem is the signature of `toString` rather than `formattedWrite` after the long journey in Phobos.

Reply:
Hmm, I see, depending on `sink`, the call to `formattedWrite` might be @safe, right?

Reply:
Maybe a better signature would be: `void toString()(scope void delegate(const(char)[]) sink)`. Templated with an empty template paramter list. That way, it is clear what type of argument the function expects.

Reply:
> Hmm, I see, depending on sink, the call to formattedWrite might be @safe, right?

I guess so.

> Maybe a better signature would be: `void toString()(scope void delegate(const(char)[]) sink)`.

Using different signature from recommended in std.format makes users confused and changing the signatures for `formatValue` is out of scope of this request.

Reply:
I'm sorry but I don't get your point. Without this PR, we have a non-templated function with the signature `void toString(scope void delegate(const(char)[]) sink)`. This function is `system` but it specifically accepts only delegate arguments, whereas with this PR it can accept any type of argument. What I am proposing is to simply template `toString` with an empty template parameter list. That way, `@safe` can be inferred but the signature remains mostly the same.

Reply:
I'm sorry, I used a wrong signature in the comment.
Please replace
> `void toString()(scope void delegate(const(char)[]) sink) @system`

with `void toString()(scope void delegate(const(char)[]) @system sink)`.
That is, even if `toString` is templated, the argument `sink` only accepts system delegates.

In addition to that, as I linked in the first comment, the [official document of std.format.formatValue](https://dlang.org/phobos/std_format.html#.formatValue) says:

> `toString` should have one of the following signatures:
> ```dlang
> void toString(W)(ref W w, scope const ref FormatSpec fmt)
> void toString(W)(ref W w)
> string toString();
> ```

Therefore using `void toString(W)(ref W w)` is reasonable.